diff -upNr pymol-open-source-2.5.0.orig/modules/pymol/importing.py pymol-open-source-2.5.0.patched/modules/pymol/importing.py
--- pymol-open-source-2.5.0.orig/modules/pymol/importing.py	2021-05-10 20:38:37.000000000 +0200
+++ pymol-open-source-2.5.0.patched/modules/pymol/importing.py	2022-01-27 21:27:57.683853866 +0100
@@ -1113,36 +1113,36 @@ PYMOL API
         return r
 
     fetchHosts = {
-        "pdb"  : "https://ftp.wwpdb.org/pub/pdb",
-        "pdbe" : "https://ftp.ebi.ac.uk/pub/databases/pdb",
+        "pdb"  : "http://ftp.wwpdb.org/pub/pdb",
+        "pdbe" : "http://ftp.ebi.ac.uk/pub/databases/pdb",
         "pdbj" : "http://ftp.pdbj.org/pub/pdb",
     }
 
     hostPaths = {
-        "mmtf" : "https://mmtf.rcsb.org/v1.0/full/{code}.mmtf.gz",
+        "mmtf" : "http://mmtf.rcsb.org/v1.0/full/{code}.mmtf.gz",
         "bio"  : [
-            "https://files.rcsb.org/download/{code}.{type}.gz",
+            "http://files.rcsb.org/download/{code}.{type}.gz",
             "/data/biounit/coordinates/divided/{mid}/{code}.{type}.gz",
         ],
         "pdb"  : [
-            "https://files.rcsb.org/download/{code}.{type}.gz",
+            "http://files.rcsb.org/download/{code}.{type}.gz",
             "/data/structures/divided/pdb/{mid}/pdb{code}.ent.gz",
         ],
         "cif"  : [
-            "https://files.rcsb.org/download/{code}.{type}.gz",
+            "http://files.rcsb.org/download/{code}.{type}.gz",
             "/data/structures/divided/mmCIF/{mid}/{code}.cif.gz",
-            "https://ftp-versioned.wwpdb.org/pdb_versioned/views/latest/coordinates/mmcif/{mid}/pdb_{code:0>8}/pdb_{code:0>8}_xyz.cif.gz",
+            "http://ftp-versioned.wwpdb.org/pdb_versioned/views/latest/coordinates/mmcif/{mid}/pdb_{code:0>8}/pdb_{code:0>8}_xyz.cif.gz",
         ],
-        "2fofc" : "https://www.ebi.ac.uk/pdbe/coordinates/files/{code}.ccp4",
-        "fofc" : "https://www.ebi.ac.uk/pdbe/coordinates/files/{code}_diff.ccp4",
+        "2fofc" : "http://www.ebi.ac.uk/pdbe/coordinates/files/{code}.ccp4",
+        "fofc" : "http://www.ebi.ac.uk/pdbe/coordinates/files/{code}_diff.ccp4",
         "pubchem": [
-            "https://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?{type}={code}&disopt=3DSaveSDF",
-            "https://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?{type}={code}&disopt=SaveSDF",
+            "http://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?{type}={code}&disopt=3DSaveSDF",
+            "http://pubchem.ncbi.nlm.nih.gov/summary/summary.cgi?{type}={code}&disopt=SaveSDF",
         ],
         "emd": "/../emdb/structures/EMD-{code}/map/emd_{code}.map.gz",
         "cc": [
-            "https://files.rcsb.org/ligands/download/{code}.cif",
-            "https://ftp.ebi.ac.uk/pub/databases/msd/pdbechem_v2/{code:.1}/{code}/{code}.cif",
+            "http://files.rcsb.org/ligands/download/{code}.cif",
+            "http://ftp.ebi.ac.uk/pub/databases/msd/pdbechem_v2/{code:.1}/{code}/{code}.cif",
         ],
     }
 
